### PR TITLE
Adding TableView section

### PR DESCRIPTION
**Pull request by e8johan**
Adding a section on TableView in the modelview chapter. Also updated the qtcpp chapter to explain that QAbstractTableModel can be used with QML and how it differs to the QAbstractListModel used in the examples.

### CODE REVIEW

File: docs/modelview/modelview.rst [hunk 1]
```diff
@@ -384,7 +430,6 @@ Having more delegates sacrifices memory for a smoother experience and slightly m
 
 To remedy the two later issues, it is recommended to use ``Loader`` elements. These can be used to instantiate additional elements when they are needed. For instance, an expanding delegate may use a ``Loader`` to postpone the instantiation of its detailed view until it is needed. For the same reason, it is good to keep the amount of JavaScript to a minimum in each delegate. It is better to let them call complex pieced of JavaScript that resides outside each delegate. This reduces the time spent compiling JavaScript each time a delegate is created.
```
Comment (richardmg) [inline docs/modelview/modelview.rst]: later or latter?

**Comment by richardmg** [comment]
This looks good to me! Nicely written. I wonder though if you have considered writing a section/side note about the DelegateChooser? It's currently a labs project, though...

File: docs/modelview/modelview.rst [hunk 2]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
 
+Table Models
+------------
+
+All views discussed until now present an array of items one way or another. Even the ``GridView`` expects the model to provide a one dimensional list of items. For two dimensional tables of data you need to use the ``TableView`` element.
+
+Conceptually, the ``TableView`` works just as all other views. It combines a ``model`` with a ``delegate`` to form a grid. If given a list type model it displays a single column, making it very similar to the ``ListView`` element. However, for table style models, it becomes more powerful. 
+
+In the example below, we setup a simple ``TableView`` with a custom model exposed from C++. At the moment, it is not possible to create table style models directly from QML, but in the 'Qt and C++' chapter the concept is explained. The running example is shown in the figure below.
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: set up
(setup is a noun, not a verb)

Comment (ec1oud) [reply docs/modelview/modelview.rst]: also this phrase "table style models" is unclear... someone might think it has something to do with visual styling.
"table-style models" or "table-oriented models" perhaps

File: docs/modelview/modelview.rst [hunk 3]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
+
+.. figure:: assets/tableview.png
+
+Before we can use the ``TableView`` element, we must make sure that the ``2.12`` version of ``QtQuick`` is imported. After that, we can set it up. In this example below, we set the ``rowSpacing`` and ``columnSpacing`` to control the horizontal and vertical gaps between delegates. The rest of the properties are setup as for any other type of view.
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: are set up

File: docs/modelview/modelview.rst [hunk 4]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
+
+.. literalinclude:: src/tableview/main.qml
+    :start-after: M1>>
+    :end-before: <<M1
+
+.. literalinclude:: src/tableview/main.qml
+    :start-after: M2>>
+    :end-before: <<M2
+    
+The delegate itself can carry an implicit size through the ``implicitWidth`` and ``implicitHeight``. This is what we do in the example below. The actual data contents, i.e. ``display``, depends on what the model exposes.
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: i.e. the string returned from the model's ``display`` role

File: docs/modelview/modelview.rst [hunk 5]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
+    :end-before: <<M3
+
+It is possible to provide delegates with different sizes depending on the model contents, e.g.::
+
+    GreenBox {
+        implicitHeight: (1+row)*10
+        // ...
+    }
+
+Notice that both the width and the height must be greater than one.
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: greater than zero

File: docs/modelview/modelview.rst [hunk 6]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
+
+When providing an implicit size from the delegate, the highest delegate of each row and the widest delegate of each column controls the size. This can create interesting behaviour if the width of items depend on the row, or if the height depends on the column. This is because not all delegates are instantiated at all times, so the width of a column might change as the user scrolls through the table.
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: s/highest/tallest/

File: docs/modelview/modelview.rst [hunk 7]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
...
+
+To avoid the issues with specifying column widths and row heights using implicit delegate sizes, you can provide functions that calculates these sizes. This is done using the ``columnWidthProvider`` and ``rowHeightProvider`` . These functions return the size of the width and row respectively as shown below::
```
Comment (ec1oud) [inline docs/modelview/modelview.rst]: functions that calculate

File: docs/modelview/modelview.rst (hunk 1 above)
Comment (ec1oud) [reply docs/modelview/modelview.rst]: latter

Putting too many Loaders into delegates is one thing that made the Controls 1 TableView perform badly.  I would do it only when a delegate is so big that it slows down flicking, or the delegate depends on remotely-fetched data, has media to render, or something like that; but not as a matter of course.  (this comment is just a comment, not for direct pasting ;-)

File: docs/modelview/src/tableview/main.qml [hunk 8]
```diff
@@ -0,0 +1,48 @@
+// M1>>
+import QtQuick 2.12
+// <<M1
+import QtQuick.Window 2.0
```
Comment (ec1oud) [inline docs/modelview/src/tableview/main.qml]: 2.12
you can use x.12 for every import, where x is the module version.  (In Qt 6 we might make it even more consistent, i.e. maybe every module will have version 6.y)

File: docs/modelview/src/tableview/tablemodel.cpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+int TableModel::columnCount(const QModelIndex &) const
+{
+    return 10;
+}
+
+QVariant TableModel::data(const QModelIndex &index, int role) const
+{
+    if (role == Qt::DisplayRole)
+    {
```
Comment (ec1oud) [inline docs/modelview/src/tableview/tablemodel.cpp]: no need for braces (but if you like braces that much, you should put them around the else block too, to be consistent)

File: docs/modelview/modelview.rst (hunk 1 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Good points. I've updated accordingly.

**Comment by e8johan** [comment]
> This looks good to me! Nicely written. I wonder though if you have considered writing a section/side note about the DelegateChooser? It's currently a labs project, though...

Thanks :-)

I'm adding DelegateChooser to my backlog of things that should be incorporated.

File: docs/modelview/modelview.rst [hunk 10]
```diff
@@ -99,7 +99,7 @@ Repeaters work well for limited and static sets of data, but in the real world,
 .. figure:: assets/automatic/gridview-basic.png
     :scale: 50%
 
-The two elements are similar in their usage. Thus, we will begin with the ``ListView`` and then describe the ``GridView`` with the former as the starting point of the comparison.
+The two elements are similar in their usage. Thus, we will begin with the ``ListView`` and then describe the ``GridView`` with the former as the starting point of the comparison. Notice that the ``GridView`` places a list of items into a two dimenstional grid. If you want to show a table of data you need to use the ``TableView`` which is described in the `Table Models`_ section.
```
[the remark this answers is not in the document]
Comment (e8johan) [reply docs/modelview/modelview.rst]: Taken!

File: docs/modelview/modelview.rst [hunk 11]
```diff
@@ -324,6 +324,52 @@ Another detail specific to ``PathView`` worth noticing is the usage of the attac
 
 When transforming images or other complex elements on in ``PathView``, a performance optimization trick that is common to use is to bind the ``smooth`` property of the ``Image`` element to the attached property ``PathView.view.moving``. This means that the images are less pretty while moving but smoothly transformed when stationary. There is no point spending processing power on smooth scaling when the view is in motion, as the user will not be able to see this anyway.
 
+Table Models
+------------
+
+All views discussed until now present an array of items one way or another. Even the ``GridView`` expects the model to provide a one dimensional list of items. For two dimensional tables of data you need to use the ``TableView`` element.
+
+Conceptually, the ``TableView`` works just as all other views. It combines a ``model`` with a ``delegate`` to form a grid. If given a list type model it displays a single column, making it very similar to the ``ListView`` element. However, for table style models, it becomes more powerful.
```
[the remark this answers is not in the document]
Comment (e8johan) [reply docs/modelview/modelview.rst]: Taken!

File: docs/modelview/modelview.rst (hunk 2 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Good points (and thanks for finally clarifying setup vs set up :-) ).

File: docs/modelview/modelview.rst (hunk 3 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Fixed!

File: docs/modelview/modelview.rst (hunk 4 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Fixed

File: docs/modelview/modelview.rst (hunk 5 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Fixed

File: docs/modelview/modelview.rst (hunk 7 above)
Comment (e8johan) [reply docs/modelview/modelview.rst]: Fixed

File: docs/modelview/src/tableview/tablemodel.cpp (hunk 9 above)
Comment (e8johan) [reply docs/modelview/src/tableview/tablemodel.cpp]: Let's drop the braces :-)

File: docs/modelview/src/tableview/main.qml (hunk 8 above)
Comment (e8johan) [reply docs/modelview/src/tableview/main.qml]: This is explained earlier (and it's great with the x.12 - you have my vote for a 6.12 approach). However, when discussing the general update of versions, we decided against changing all the import statements to the latest version and instead show that features are added over time.

**Comment by e8johan** [comment]
Regarding the header demo proposed by @ec1oud , I'll add that to the backlog as well - but thanks for the input. Can you tell me when you might have made this easier, as it might be better to wait for the proper solution if it is due soon.

**Comment by ec1oud** [comment]
> Can you tell me when you might have made this easier
I hope we get something done in time for 5.13.  But there are a lot of things like that to do now.

**Pull request merged**